compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExposedVisibilityChecker.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.fir.*
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirDiagnosticFactory3
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.FirTypeRef
+import org.jetbrains.kotlin.fir.types.coneTypeSafe
+
+object FirExposedVisibilityChecker : FirDeclarationChecker<FirMemberDeclaration>() {
+    override fun check(declaration: FirMemberDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
+        when (declaration) {
+            is FirTypeAlias -> checkTypeAlias(declaration, context, reporter)
+            is FirProperty -> checkProperty(declaration, context, reporter)
+            is FirFunction<*> -> checkFunction(declaration, context, reporter)
+        }
+    }
+
+    private fun checkTypeAlias(declaration: FirTypeAlias, context: CheckerContext, reporter: DiagnosticReporter) {
+        val expandedType = declaration.expandedConeType
+        val typeAliasVisibility = declaration.firEffectiveVisibility(declaration.session)
```

Comment:
Please use just `.effectiveVisibility` here and below. It works more correctly for local classes.

Reply:
Now using the combination of `.effectiveVisibility` and `.firEffectiveVisibility(session) `